### DISCUSSION TITLE
Use of core directly from C++

### DISCUSSION THREAD

**Discussion by rschwarz** [metab0t/PyOptInterface#25] [General]
I understand that a common core and the different solver interfaces are all implemented in C++, and then wrapped as a Python library for convenience/ergonomics.

But in principle, the core should be usable by itself, as a C++ library, right? Has this been done before? Are there some gotchas to consider?

It would still bring the benefit of a high-performance modeling layer that supports multiple solvers transparently, so I think it can be quite valuable.

**Comment by metab0t** [comment]
Hi, Schwarz.

I have intentionally separated the responsibility between the C++ part and the Python part. The C++ part is definitely usable independently.

For example, if you want to use Gurobi with C++, just grab the following files, and you can build a standalone static library to link with your application:
- `include/container.hpp`
- `include/core.hpp`
- `include/dylib.hpp`
- `include/solver_common.hpp`
- `lib/core.cpp`
- `include/gurobi_model.hpp`
- `lib/gurobi_model.cpp`

By the way, remember to adjust the include path of `gurobi_c.h` at the beginning of `include/gurobi_model.hpp` because the header file is vendored in PyOptInterface repo.

For HiGHS and other solvers, you can include the corresponding header and implementation files, respectively. As we use dynamic loading of shared libraries, you can always build these files independently with a working C++ compiler and do not need binary of solvers.

We all know that packaging and distributing C++ library correctly is not an easy task, so I think it is a better idea to grab the sources and build it as a standalone library for interested users.

**Reply by rschwarz** [reply]
OK, so if I wanted to try this, it's probably best to "vendor in" the source code in my own project.

**Reply by metab0t** [reply]
Yes, you can include these files in your project and write a CMakeLists.txt to build a static library. Cross-platform dependency management in C++ is quite difficult, and I choose to "vendor in" carefully selected dependencies in PyOptInterface.

I forget to mention that there are also two external dependencies: [fmtlib](https://github.com/fmtlib/fmt) and [martinus/unordered_dense](https://github.com/martinus/unordered_dense). You might already depend on `fmtlib` as a good alternative for `printf`, and the latter is a single header-only file.
